Add docker-compose for quick dev env

LoveIsGrief opened the pull request:
#3116 (gitpod support) made me realize it'd be easier to work with something I know, so here's a PR to help newbies like me to setup a new environment without having to deal with installing LAMP on the host machine.

I added the information to `CONTRIBUTING.md` because I don't have access to the wiki.

It works without a hitch in my linux env + PHPStorm + XDebug.

Devs can now `docker-compose up -d` and access http://localhost:8080 with user:passwd `ampache:ampache`. They can shove some media files into `docker/volumes/media` and create a catalog there which works.

With a debugger of their choice, they can listen on host port `9003`, create a `docker-compose.override.yml` (see `CONTRIBUTING.md`) and `docker-compose up -d ampache` to get callbacks from XDebug.

lachlan-00 commented:
cool, i'm going to be away for a week from tomorrow but when i come back i'll set myself up this way and test it all out.

I'll be back about the 24/25th and will merge this when i've tested how it works and can update the wiki too

LoveIsGrief commented:
I hope it'll work and look forward to the feedback.

Happy holidays! :bell:

LoveIsGrief commented:
@kuzi-moto we should probably have the discussion about docker in here. I can always drop the commits from #3124 that have the docker stuff in them.

I had a look at the https://github.com/ampache/ampache-docker/blob/master/Dockerfile and keeping it in a separate repo has the problems, imo:

 - Builds aren't triggered through DockerHub automatically when this repo is updated/pushed (probably manually done or with an external service to trigger Dockerhub)
 - It's a monolith image with all services in one --> adding separate services e.g redis or nginx requires rebuilding the entire image
 - Using the repo in here either means a submodule (either in here of over there) which comes with its own set of complications
   * Not everybody knows about submodules, so there will be separate documentation needed that will have to be easy to find
   * Modifying the `Dockerfile` / docker setup will require the following steps before merging it into this repo:
      - Forking ampache/docker
      - ampache/ampache: pointing the submodule to a forked repo
      - ampache/docker: Getting a PR for merged there
      - ampache/ampache: pointing to the `master` tip of `ampache/docker`

It's definitely the latter reason why I put all the docker stuff in here. I didn't it up for Dockerhub since I was/am purely focused on getting a developer environment up and running. I simply couldn't find any documentation on how to do that and have never used PHP before.

lachlan-00 commented:
back,

why is the website argument used for email? that's a bit of a bug.

lachlan-00 commented on docker-compose.override-template.yml:
```diff
@@ -0,0 +1,25 @@
+version: "3"
```
move this into templates folder `docs/templates/docker-compose.override-template.yml`

lachlan-00 commented on .gitignore:
```diff
@@ -50,3 +50,8 @@ web.config
 *.v11.suo
 *~
 config/ampache.cfg.php.backup
+
```
put this just in the file as normal
![image](https://user-images.githubusercontent.com/1305249/147995500-9ab3af1b-d0fc-4e7b-9431-9b8d550cd2b7.png)

lachlan-00 commented on CONTRIBUTING.md:
```diff
@@ -7,7 +7,41 @@ submitting bug reports and feature requests or writing code which can be incorpo
 Anyone can take part in our community and it there are no rules or requirements stopping you from joining.
 Hopefully this document will help you make the jump!
 
-Please read [Development section](https://github.com/ampache/ampache/wiki#development).
```
duplicate the updated file in `.github/CONTRIBUTING.md`

LoveIsGrief replied:
Not for this PR, but maybe a symlink would work too. I copied it nonetheless.

lachlan-00 commented on src/Module/Cli/AdminAddUserCommand.php:
```diff
@@ -55,11 +55,26 @@ public function execute(
         $values     = $this->values();
         $interactor = $this->io();
 
+        $email = $values['website'];
```
is this a bug you found? i'd keep it simple with checking email and user separately
![image](https://user-images.githubusercontent.com/1305249/147995591-d2dd1557-c768-4979-a6b2-ba32a605f746.png)

LoveIsGrief replied:
Yeah, I was able to add two users with the same email and username. I'm not going to touch migrations in here, but the email and username columns should probably be `UNIQUE`.

lachlan-00 commented:
update from develop as well and when we get this all done i'll just squash it down into one commit

LoveIsGrief commented:
It should be ready now @lachlan-00

lachlan-00 commented:
looks good now, i'll run it up here and if i can handle it then everyone else can :)

lachlan-00 commented on docker-compose.override-template.yml:
```diff
@@ -1,25 +0,0 @@
-version: "3"
```
might not have added this in to the pull?

LoveIsGrief commented:
It's renamed and moved in the docs/templates/docker-compose.override.yml

I figured `override-template` wasn't necessary since it's in the templates folder anyway.

lachlan-00 commented:
the web server should go to the root instead of /public as well. will probably create a few errors if not using the right webroot.

not sure how it's configured in nginx but will try again tomorrow

LoveIsGrief commented:
Hmm, I'm not sure what you mean. The root is /var/www/html and it only redirects to /public when when matching exactly "/". Even /index.php states that the root ampache folder is /public.
The repo root is mounted at /var/www/html in the nginx and ampache services.

lachlan-00 commented:
when i go to localhost:8080 i get redirected to /public
![image](https://user-images.githubusercontent.com/1305249/148475850-39ecd9c1-4bbf-4b6f-b665-706d3f634b18.png)

The public folder is where the document root for the server should be

LoveIsGrief commented:
OK, I see. I misunderstood the message in `/index.php`. Your issue was actually being caused by the `ampache` service not being available (502 means nginx can't reach the upstream host), due to the db not being ready during ampache's setup script run. `docker-compose start ampache` would've resolved it.

I added a `restart: on-failure` for the `ampache` service to make manual restarting unnecessary, updated the document root and removed `web_path = "/public";` from the config to remove `/public/` from URLs.

lachlan-00 commented:
i get it now, my docker lock was set to php7.4 which was stopping the service completing

LoveIsGrief commented:
Is there anything else I need to do?

lachlan-00 commented:
Still not working well on this pc but i'm going to clean up my other laptop and try to run it from there. i think i screwed around too much with the docker on this thing so i think it's me whos causing the problems rather than what's here now

lachlan-00 commented:
i'm looking at making this a bit simpler, the nginx stack is just overly complicated for me (mostly because i don't use nginx or docker regularly) and i have not been able to get it working at all on my PC.

https://github.com/ampache/ampache/tree/localdevenv

@LoveIsGrief and @kuzi-moto i've added this branch for a bit of a review compared to this pull using apache2 instead

i've tried to keep this simple by running a shell script with the perms and config. from there build and compose up should make it pretty simple.
![image](https://user-images.githubusercontent.com/1305249/150052500-1c1fdfe7-cd0a-49b8-8c83-f1345df62410.png)

The only thing i'm not sure about is xdebug which i've enabled as a php extension but not something i actually use either

LoveIsGrief commented:
I'm not sure why it didn't work for you, but just to be sure it wasn't a "it works for me" problem, I tested my branch in clean VMs running Arch Linux, OpenSuse 15, and Ubuntu 20.04. They all work https://github.com/LoveIsGrief/ampache-dev-docker-test . Even tested it on my local NixOS and it works too. What errors were you running into?

`nginx` is in a separate container in order to allow anybody to switch out the webserver for whatever they like: apache, httpd, caddy, etc. `nginx.conf` was also taken from the [wiki](https://github.com/ampache/ampache/wiki/Installation#nginx).

The reason I didn't add any scripts to run before running docker was because imo, somebody new to the project should just be able to run `docker-compose up` or a single command and be up and running. Additionally, no further steps e.g creating a user, configuring the DB access, etc. should be necessary.

lachlan-00 commented:
for me the nginx site isn't able to connect to anything so the ampache-fpm and db are running okay as far as i can tell

I haven't been able to site down even on another laptop for a while so it's one of thing things i haven't really worked out why yet.

LoveIsGrief commented:
I'm not using ampache anymore and don't think I'll be touching PHP again, so I'm closing this
